cmd/pebble/cmd_logs.go

```diff
@@ -15,27 +15,129 @@
 package main
 
 import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
 	"github.com/jessevdk/go-flags"
+
+	"github.com/canonical/pebble/client"
 )
 
 type cmdLogs struct {
 	clientMixin
+	Follow     bool   `short:"f" long:"follow"`
+	Format     string `long:"format"`
+	NumLogs    string `short:"n" long:"number"`
 	Positional struct {
 		Services []string `positional-arg-name:"<service>"`
 	} `positional-args:"yes"`
 }
 
-var logsDescs = map[string]string{}
+var logsDescs = map[string]string{
+	"follow": "Follow (tail) logs for given services until Ctrl-C pressed.",
+	"format": `Output format: "text" (default) or "json" (JSON lines).`,
+	"number": "Number of logs to show (before following); defaults to 10.\nIf 'all', show all buffered logs.",
+}
 
-var shortLogsHelp = "Retrieve logs for service"
+var shortLogsHelp = "Fetch service logs"
 var longLogsHelp = `
-The logs command fetches logs of the given services and displays them in
-chronological order.
+The logs command fetches buffered logs from the given services (or all services
+if none are specified) and displays them in chronological order.
 `
 
 func (cmd *cmdLogs) Execute(args []string) error {
+	var numLogs int
+	switch cmd.NumLogs {
+	case "":
+		numLogs = 10
+	case "all":
+		numLogs = -1
+	default:
+		var err error
+		numLogs, err = strconv.Atoi(cmd.NumLogs)
+		if err != nil {
+			return fmt.Errorf(`expected n to be a non-negative integer or "all", not %q`, cmd.NumLogs)
+		}
+	}
+
+	var writeLog client.WriteLogFunc
+	switch cmd.Format {
+	case "", "text":
+		writeLog = writeLogText
+	case "json":
+		writeLog = writeLogJSON
```

Comment:
I think we should also just add "raw", which io.Copy the message to the corresponding output stream.

Reply:
Done.